aptos-move/framework/aptos-framework/sources/account.move

```diff
@@ -249,6 +255,82 @@ module aptos_framework::account {
+        if (scheme == ED25519_SCHEME) {
+            let pk = ed25519::new_unvalidated_public_key_from_bytes(public_key_bytes);
+            let sig = ed25519::new_signature_from_bytes(signature);
+            assert!(ed25519::signature_verify_strict_t(&sig, &pk, *challenge), std::error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+            ed25519::unvalidated_public_key_to_authentication_key(&pk)
+        } else {
+            let pk = multi_ed25519::new_unvalidated_public_key_from_bytes(public_key_bytes);
+            let sig = multi_ed25519::new_signature_from_bytes(signature);
+            assert!(multi_ed25519::signature_verify_strict_t(&sig, &pk, *challenge), std::error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+            multi_ed25519::unvalidated_public_key_to_authentication_key(&pk)
+        }
+    }
+
+    /// Generic authentication key rotation function that allows the user to rotate their authentication key from any scheme to any scheme.
+    /// To authorize the rotation, a signature by the current private key on a valid RotationProofChallenge (`cap_rotate_key`)
+    /// demonstrates that the user intends to and has the capability to rotate the authentication key. A signature by the new
+    /// private key on a valid RotationProofChallenge (`cap_update_table`) verifies that the user has the capability to update the
+    /// value at key `auth_key` on the `OriginatingAddress` table. `from_scheme` refers to the scheme of the `from_public_key` and
+    /// `to_scheme` refers to the scheme of the `to_public_key`. A scheme of 0 refers to an Ed25519 key and a scheme of 1 refers to
+    /// Multi-Ed25519 keys.
+    public entry fun rotate_authentication_key(
+        account: &signer,
+        from_scheme: u8,
+        from_public_key_bytes: vector<u8>,
+        to_scheme: u8,
+        to_public_key_bytes: vector<u8>,
+        cap_rotate_key: vector<u8>,
+        cap_update_table: vector<u8>,
+    ) acquires Account, OriginatingAddress {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+        assert!((from_scheme == ED25519_SCHEME || from_scheme == MULTI_ED25519_SCHEME), EINVALID_SCHEME);
+        assert!((to_scheme == ED25519_SCHEME || to_scheme == MULTI_ED25519_SCHEME), EINVALID_SCHEME);
+
+        let account_resource = borrow_global_mut<Account>(addr);
+
+        // verify the public key matches the current authentication key
+        if (from_scheme == ED25519_SCHEME) {
+            let from_pk = ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
+            let from_auth_key = ed25519::unvalidated_public_key_to_authentication_key(&from_pk);
+            assert!(account_resource.authentication_key == from_auth_key, error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
+        } else {
+            let from_pk = multi_ed25519::new_unvalidated_public_key_from_bytes(from_public_key_bytes);
```

Comment:
@chloeqjz @alinush can we also validate that the multisig public keys have at most 32 ed25519 keys in them? It is better to be safe

Reply:
I think alin already checked this in the rust implementation: https://github.com/aptos-labs/aptos-core/blob/main/crates/aptos-crypto/src/multi_ed25519.rs#L64

Reply:
yes, rust has that. but that restriction doesn't limit users from sending a `to_public_key_bytes ` with more than 32 keys here, right? That means the user might be able to rotate with a multisig key that is invalid and then the account cannot be accessed anymore. Is this the case?

Reply:
ooh so I think this function 
`verify_key_rotation_signature_and_get_auth_key(to_scheme, to_public_key_bytes, cap_update_table, &challenge)` (https://github.com/aptos-labs/aptos-core/blob/2e9b216d0e184af6d2f8db3d78785fc9697d671f/aptos-move/framework/aptos-framework/sources/account.move#L267) calls `native_signature_verify_strict ` (https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/src/natives/cryptography/multi_ed25519.rs#L58). and `native_signature_verify_strict ` calls `multi_ed25519::MultiEd25519PublicKey::try_from(pubkey.as_slice()) ` and that function calls `check_and_get_threshold` where they have the following code `
    if num_of_keys == 0 || num_of_keys > MAX_NUM_OF_KEYS || threshold_num_of_bytes != 1 {
        Err(CryptoMaterialError::WrongLengthError)
    } ` to check the length. I think you're right that they may be able to send a invalid `to_public_key_bytes` but this should fail in the signature verification step

Reply:
Right... you both have good points:
1. @jjleng, we could detect this earlier in `new_unvalidated_public_key_from_bytes`
2. @chloeqjz, we do eventually detect this in the native signature verification.

@chloeqjz, can we nonetheless implement the check you Slack'd me with in `new_unvalidated_public_key_from_bytes`, just to make sure we error as early as possible?
Specifically,
```
assert!(vector::length(&bytes) / INDIVIDUAL_PUBLIC_KEY_NUM_BYTES <= MAX_NUMBER_OF_PUBLIC_KEYS, error::invalid_argument(E_WRONG_PUBKEY_SIZE));
```
and define `MAX_NUMBER_OF_PUBLIC_KEYS` to be 32!

Reply:
In addition, can you also return false earlier in our `native_public_key_validate` function that implements MultiEd25519 public key validation? Specifically, at line 33 [here](https://github.com/aptos-labs/aptos-core/blob/9366413f678e55f2babd3e7da192d0ecce58ae40/aptos-move/framework/src/natives/cryptography/multi_ed25519.rs#L33), check that the # of sub public keys is not > 32:

```
    if num_sub_pks > multi_ed25519::MAX_NUM_OF_KEYS {
        return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)]));
    }
```

You will need to make MAX_NUM_OF_KEYS public in the MultiEd25519 file from `aptos-crypto/`